graphene_django/fields.py

```diff
@@ -103,39 +102,44 @@ def resolve_connection(cls, connection, default_manager, args, iterable):
 
     @classmethod
     def connection_resolver(
-        cls,
-        resolver,
-        connection,
-        default_manager,
-        max_limit,
-        enforce_first_or_last,
-        root,
-        info,
-        **args
-    ):
-        first = args.get("first")
-        last = args.get("last")
-
-        if enforce_first_or_last:
-            assert first or last, (
-                "You must provide a `first` or `last` value to properly paginate the `{}` connection."
-            ).format(info.field_name)
-
-        if max_limit:
-            if first:
-                assert first <= max_limit, (
-                    "Requesting {} records on the `{}` connection exceeds the `first` limit of {} records."
-                ).format(first, info.field_name, max_limit)
-                args["first"] = min(first, max_limit)
-
-            if last:
-                assert last <= max_limit, (
-                    "Requesting {} records on the `{}` connection exceeds the `last` limit of {} records."
-                ).format(last, info.field_name, max_limit)
-                args["last"] = min(last, max_limit)
-
-        iterable = resolver(root, info, **args)
-        on_resolve = partial(cls.resolve_connection, connection, default_manager, args)
+            cls,
+            resolver,
+            connection,
+            default_manager,
+            max_limit,
+            enforce_first_or_last,
+            root,
+            info,
+            **kwargs):
+        # pylint: disable=R0913,W0221
+
+        first = kwargs.get("first")
+        last = kwargs.get("last")
+        if not (first is None or first > 0):
+            raise ValueError(
+                "`first` argument must be positive, got `{first}`".format(**locals()))
+        if not (last is None or last > 0):
+            raise ValueError(
+                "`last` argument must be positive, got `{last}`".format(**locals()))
+        if enforce_first_or_last and not (first or last):
+            raise ValueError(
+                "You must provide a `first` or `last` value "
+                "to properly paginate the `{info.field_name}` connection.".format(**locals()))
+
+        if not max_limit:
+            pass
+        elif first is None and last is None:
+            kwargs['first'] = max_limit
+        else:
+            count = min(i for i in (first, last) if i)
```

Comment:
Is it even valid to have a first and a last? I'm not sure I fully understand what behaviour providing `first` and `last` should result in. Could you give an example?

Reply:
By current `graphql_relay` implementation,
`last` means last N items in the selection,
and when use `last` with `first`,  `last` means last N items in first N items.